common/src/main/java/io/pravega/common/util/CompositeByteArraySegment.java

```diff
@@ -112,14 +113,126 @@ private CompositeByteArraySegment(Object[] arrays, int arraySize, int startOffse
-        byte[] array = getArray(getArrayId(offset), true); // Need to allocate array if not allocated yet.
-        array[getArrayOffset(offset)] = value;
+        ByteBuffer bb = getBuffer(getBufferId(offset), true); // Need to allocate array if not allocated yet.
+        bb.array()[getBufferOffset(offset)] = value;
+    }
+
+    @Override
+    public void setShort(int offset, short value) {
+        Exceptions.checkArrayRange(offset, Short.BYTES, this.length, "index", "length");
+
+        int bufferId = getBufferId(offset);
+        ByteBuffer bb = getBuffer(bufferId, true);
+        int bufferOffset = getBufferOffset(offset);
+        try {
+            bb.putShort(bufferOffset, value);
+        } catch (IndexOutOfBoundsException | BufferOverflowException ex) {
+            bb.array()[bufferOffset] = (byte) (value >>> 8 & 255);
+            getBuffer(bufferId + 1, true).array()[0] = (byte) (value & 255);
+        }
+    }
+
+    @Override
+    public void setInt(int offset, int value) {
+        Exceptions.checkArrayRange(offset, Integer.BYTES, this.length, "index", "length");
+
+        int bufferId = getBufferId(offset);
+        ByteBuffer bb = getBuffer(bufferId, true);
+        int bufferOffset = getBufferOffset(offset);
+
+        try {
+            bb.putInt(bufferOffset, value);
+            return;
+        } catch (IndexOutOfBoundsException | BufferOverflowException ex) {
+            // Intentionally left blank.
+        }
+
+        int diff = bb.remaining() - bufferOffset;
+        if (diff == 3) {
+            // We can only fit 3 bytes. Write as 1 short + 2 bytes.
+            bb.putShort(bufferOffset, (short) (value >>> 16));
+            bb.array()[bufferOffset + Short.BYTES] = (byte) (value >>> 8);
+            getBuffer(bufferId + 1, true).array()[0] = (byte) value;
+        } else if (diff == 2) {
+            // We can only fit 2 bytes. Write as 2 shorts
+            bb.putShort(bufferOffset, (short) (value >>> 16));
+            getBuffer(bufferId + 1, true).putShort(0, (short) value);
+        } else {
+            // We can only fit 1 byte. Write as 2 bytes + 1 short.
+            bb.array()[bufferOffset] = (byte) (value >>> 24);
+            bb = getBuffer(bufferId + 1, true);
+            bb.array()[0] = (byte) (value >>> 16);
+            bb.putShort(1, (short) value);
+        }
+    }
+
+    @Override
+    public void setLong(int offset, long value) {
+        Exceptions.checkArrayRange(offset, Long.BYTES, this.length, "index", "length");
+
+        int bufferId = getBufferId(offset);
+        ByteBuffer bb = getBuffer(bufferId, true);
+        int bufferOffset = getBufferOffset(offset);
+
+        try {
+            bb.putLong(bufferOffset, value);
+            return;
+        } catch (IndexOutOfBoundsException | BufferOverflowException ex) {
+            // Intentionally left blank.
+        }
+
+        int diff = bb.remaining() - bufferOffset;
+        if (diff == 7) {
```

Comment:
for loops take extra CPU cycles. See my explanation above. Plus I want to break down the long into as "big" components as I can so I can write them like that - bigger components may take advantage of intrinsic operations inside these serializers.

See how `Unsafe.putLongUnaligned` works. It follows exactly the same pattern and for the same reason: it's a very low-level API call which needs to be as optimized as possible.

Reply:
If it is "clear" to the compiler that the loop is of small constant size, it can unroll it and make a computed jump. That would be more efficient than even this, and would look nice too.

Reply:
Ok, I did not notice any difference when I ran this in a tight loop, but given that the code looks nicer, I have made the change anyway.